src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+    # Normalize unorm and perform Gaussian reduction if required
+    if new
+        new = false
+        odet.unorm0[1:odet.msol] .= odet.unorm[1:odet.msol]
+    else
+        odet.unorm[1:odet.msol] .= odet.unorm[1:odet.msol] ./ odet.unorm0[1:odet.msol]
+        uratio = maximum(odet.unorm[1:odet.msol]) / minimum(odet.unorm[1:odet.msol])
+        if uratio > ctrl.ucrit || sing_flag
+            ode_fixup(intr, odet, dout, fNames, sing_flag, false)
+            new = true
+        end
+    end
+
+    return
+end
+
+function ode_fixup(intr::DconInternal, odet::OdeState, dout::DconOutput, fNames::DconFileNames, sing_flag::Bool, test::Bool)
+    # ...existing code... --> diagnose stuff would go here
+    
+    secondary = false
+    mask = trues(2, odet.msol)
+    jmax = zeros(Int, 1)
+
+    # Initial output
+    println(fNames.crit_out_unit, "Gaussian Reduction at istep = $(odet.istep), psi = $(odet.psifac), q = $q")
+    if !sing_flag
+        println(fNames.crit_out_unit, "Gaussian Reduction at istep = $(odet.istep), psi = $(odet.psifac), q = $q")
+    end
+    istate = 1
+    flag_count = 0
+
+    # Initialize fixfac
+    if !test
+        fixfac .= 0
+        for isol in 1:odet.msol
+            fixfac[isol, isol] = 1
+        end
+        # Sort unorm
+        index[1:odet.msol] .= collect(1:odet.msol)
+        index[1:intr.mpert] .= sortperm_subrange(odet.unorm, 1:intr.mpert) # in original Fortran: bubble(unorm, index, 1, mpert)
+        index[intr.mpert+1:odet.msol] .= sortperm_subrange(odet.unorm, intr.mpert+1:odet.msol) # in original Fortran: bubble(unorm, index, mpert + 1, msol)
+    end
+
+    # Triangularize primary solutions
+    mask .= true
+    for isol in 1:intr.mpert
+        ksol = index[isol]
+        mask[2, ksol] = false
+        if !test
+            # Find max location
+            absvals = abs.(u[:, ksol, 1])
+            masked = absvals .* mask[1, 1:intr.mpert]
+            kpert = argmax(masked)
+            #kpert = findmax(abs.(odet.u[:, ksol, 1]) .* (mask[1, 1:intr.mpert] .|> Int))[2] #TODO: ChatGPT suggests this might be better than the above three lines
+            mask[1, kpert] = false
+        end
+        for jsol in 1:odet.msol
+            if mask[2, jsol]
+                if !test
+                    fixfac[ksol, jsol] = -odet.u[kpert, jsol, 1] / odet.u[kpert, ksol, 1]
+                end
+                odet.u[:, jsol, :] .= odet.u[:, jsol, :] .+ odet.u[:, ksol, :] .* fixfac[ksol, jsol]
+                if !test
+                    odet.u[kpert, jsol, 1] = 0
+                end
+            end
+        end
+    end
+
+    # Triangularize secondary solutions
+    if odet.msol > intr.mpert && secondary
+        mask .= true
+        for isol in intr.mpert + 1:odet.msol
+            ksol = index[isol]
+            mask[2, ksol] = false
+            if !test
+                absvals = abs.(u[:, ksol, 2])
+                masked = absvals .* mask[1, 1:intr.mpert]
+                kpert = argmax(masked)
```

Comment:
This is a lot of unnecessary allocations inside a loop. Consider doing this in an allocation free way, or preallocating storage vectors outside the loop to use if needed.